x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/optimizer/Optimizer.java

```diff
@@ -1060,9 +1076,171 @@ private Expression propagate(And and) {
+                for (NotEquals neq : frNotEquals) {
+                    if (eq.left().semanticEquals(neq.left())) { // a = 2 OR a != ? -> TRUE
+                        return TRUE;
+                    }
+                }
+
+                Object eqValue = eq.right().fold();
+
+                // Equals OR Range
+                for (int i = 0; i < ranges.size(); i ++) { // might modify list, so use index loop
+                    Range range = ranges.get(i);
+                    if (! eq.left().semanticEquals(range.value())) {
+                        continue;
+                    }
+
+                    Integer lowerComp = range.lower().foldable() ? BinaryComparison.compare(eqValue, range.lower().fold()) : null;
+                    Integer upperComp = range.upper().foldable() ? BinaryComparison.compare(eqValue, range.upper().fold()) : null;
+
+                    if (lowerComp != null && lowerComp == 0 && !range.includeLower()) { // a = 2 OR 2 < a < ? -> 2 <= a < ?
+                        iterEq.remove(); // update range with lower equality instead
+                        ranges.set(i, new Range(range.source(), range.value(), range.lower(), true, range.upper(), range.includeUpper()));
+                        updated = true;
+                    } else if (upperComp != null && upperComp == 0 && !range.includeUpper()) { // a = 2 OR ? < a < 2 -> ? < a <= 2
+                        iterEq.remove(); // update range with upper equality instead
+                        ranges.set(i, new Range(range.source(), range.value(), range.lower(), range.includeLower(), range.upper(), true));
+                        updated = true;
+                    } else if (lowerComp != null && upperComp != null) {
+                        if (0 < lowerComp && upperComp < 0) { // a = 2 OR 1 < a < 3
+                            iterEq.remove(); // equality is superfluous
+                            updated = true;
+                        }
+                    }
+                }
+
+                // Equals OR Inequality
+                for (int i = 0; i < frInequalities.size(); i ++) {
+                    BinaryComparison bc = frInequalities.get(i);
+                    if (! eq.left().semanticEquals(bc.left())) {
+                        continue;
+                    }
+
+                    Integer comp = BinaryComparison.compare(eqValue, bc.right().fold());
+                    if (comp == null) {
+                        continue;
+                    }
+                    if (bc instanceof GreaterThan || bc instanceof GreaterThanOrEqual) {
+                        if (comp < 0) { // a = 1 OR a > 2 -> nop
+                            continue;
+                        } else if (comp == 0 && bc instanceof GreaterThan) { // a = 2 OR a > 2 -> a >= 2
+                                frInequalities.set(i, new GreaterThanOrEqual(bc.source(), bc.left(), bc.right()));
+                        } // else (0 < comp) : // a = 3 OR a > 2 -> a > 2
```

Comment:
The commented potential condition here - `0 < comp` - is not entirely true. A potential `else` branch can also happen for `bc instanceof GreaterThanOrEqual`. I don't think it's changing the logic, but just saying.

Reply:
Right, better be thorough also in comments. Fixed in both affected places.